compiler/src/iree/compiler/Codegen/Common/GPU/GPUPruneDPSOps.cpp

```diff
@@ -0,0 +1,129 @@
+  auto isPassThroughOp = [](Operation *op) -> bool {
+    return isa<linalg::PackOp, tensor::ExtractSliceOp, tensor::DimOp,
+               tensor::PadOp>(op);
+  };
+  auto isGlobalLoadOp = [](Operation *op) -> bool {
+    return isa<IREE::TensorExt::DispatchTensorLoadOp>(op);
+  };
+  // An arbitrary choice of too big.
+  auto isAllocSizeTooBig = [](Type typ) -> bool {
+    auto shapedType = cast<ShapedType>(typ);
+    int allocSize = 1;
+    for (auto dimSize : shapedType.getShape()) {
+      if (ShapedType::isDynamic(dimSize)) {
+        continue;
+      }
+      allocSize *= dimSize;
+    }
+    return allocSize >= 32;
+  };
+
+  // Iterate over all DPS ops and their inits, collecting ops to add a
+  // tensor_alloc for.
+  // There are 3 conditions of interest:
+  // 1) The result associated to the current init is unused.
+  // 2) The allocation size if appropriately small.
+  // 3) The reverse use-def chain of this op is a global load followed by ops
+  //    that do not change the address space.
+  // The 3rd condition is not strictly necessary but this is just a precation
+  // given the hacky nature of this fix.
+  SmallVector<OpOperand *> worklist;
+  funcOp.walk([&](DestinationStyleOpInterface dpsOp) {
+    for (int idx = 0; idx < dpsOp.getNumDpsInits(); ++idx) {
+      OpOperand *value = dpsOp.getDpsInitOperand(idx);
+      // If the associated result is used, do nothing.
+      if (!dpsOp->getResult(idx).use_empty()) {
+        continue;
+      }
+
+      // If the buffer allocation is too large for private space, do nothing.
+      if (isAllocSizeTooBig(value->get().getType())) {
+        continue;
+      }
+
+      // Traverse use-def chain, returning all ops that aren't subviews or
+      // slices of a global load (ops which don't change the addr space from
+      // global).
+      auto stoppingConditionFn = [&](Value a) -> bool {
+        auto op = a.getDefiningOp();
+        return !isPassThroughOp(op);
+      };
+      SetVector<Value> producers = analysisState.findValueInReverseUseDefChain(
+          value, stoppingConditionFn);
+
+      // If producer of this OpOperand is some value that isn't a subview or
+      // slice of a global load, do nothing.
+      auto arePartOfGlobalChain = [&](Value a) -> bool {
+        auto op = a.getDefiningOp();
+        return isPassThroughOp(op) || isGlobalLoadOp(op);
+      };
+      bool globalProducer = all_of(producers, arePartOfGlobalChain);
+      if (!globalProducer) {
+        continue;
+      }
+      worklist.push_back(value);
+    }
+  });
+  // Create alloc for each dps op and set it as the new in-place result.
+  for (auto value : worklist) {
+    auto dpsOp = value->getOwner();
+    PatternRewriter rewriter(context);
+    rewriter.setInsertionPoint(value->getOwner());
+    FailureOr<Value> copy = allocateTensorForShapedValue(
+        rewriter, dpsOp->getLoc(), value->get(), options, true);
+    if (failed(copy)) {
+      return signalPassFailure();
+    }
```

Comment:
Please emit an error. Otherwise, it just exit(1). The user experience would be very bad.

```suggestion
    if (failed(copy)) {
      funcOp.emitError("....");
      return signalPassFailure();
    }
```